[10.x.x] Optimize Shader Stripper

phi-lira opened the pull request:
### Purpose of this PR
 Optimize shader stripper. This PR removes GC.Allocs from CoreUtils.HasFlag and shuffles things around for better performance.

 This reduces approximately 1 min from building the template scene.

 Before:
<img width="680" alt="Screenshot 2020-08-03 at 15 36 28" src="https://user-images.githubusercontent.com/7453395/89192515-dc900400-d5a4-11ea-939f-165fec53cf76.png">
 After:
<img width="380" alt="Screenshot 2020-08-03 at 16 08 12" src="https://user-images.githubusercontent.com/7453395/89192333-86bb5c00-d5a4-11ea-9a05-952a5427d3c7.png">



 I've also tested with an artificial test by adding 10x multi_compiles to Lit.shader and building an empty cube scene. 
 The improvement seen was 20x speedup in the C# PreprocessShaders function (not 20x on the build time).
---
### Testing status

**Manual Tests**: Tested improvement by building template scene.
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

stramit commented on com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:
```diff
@@ -40,10 +50,26 @@ internal class ShaderPreprocessor : IPreprocessShaders
-        ShaderKeyword m_DeprecatedShadowsEnabled = new ShaderKeyword("_SHADOWS_ENABLED");
-        ShaderKeyword m_DeprecatedShadowsCascade = new ShaderKeyword("_SHADOWS_CASCADE");
-        ShaderKeyword m_DeprecatedLocalShadowsEnabled = new ShaderKeyword("_LOCAL_SHADOWS_ENABLED");
+        ShaderKeyword Feature00 = new ShaderKeyword("_FEATURE00");
```
Do you mean to have these features added to this?

phi-lira replied:
No, that's Draft PR, it's not ready for review it. I just opened it so I can diff easier code to make tests ;)

pbbastian commented on com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:
```diff
@@ -1,9 +1,12 @@
+//#define PROFILE_BUILD
```
Do the profiling calls actually cause that much overhead? Sounds like the markers are in a too tight loop if that is the case. If they're causing that noticeable overhead, they probably won't actually give useful profiling data.

phi-lira replied:
Actually this line should be removed as the way to set these defines should be done on the project settings.
 
This is guarded because I enable Profiler to write a file using `Profiler` API. I actually don't know if there's an overhead to have the Profiler enabled but there's a huge cost to disk writing (one of my files had 40GB on disk).

phi-lira commented:
Changes were caused by an UPM-CI commit pushing to my branch. I've merged master to remove it.
 About perforamance test, I've tested by building a scene with a single cube using Lit shader. I've added 10 multi_compiles and stripped those in the stripper. I had 20x speed increase and removed about 1GB of heap alloc caused by CoreUtils.HasFlag doing boxing, I've made further optimization but didn't manage to get updated results as Unity started soft locking after writing some profiler files after a while.
